R/Jn_crt2.R

```diff
@@ -76,55 +76,74 @@ Jn_crt2 <- function(d_est, d_sd, rho_est, rho_sd, rsq2 = 0, J = NULL,
+    criteria <- al_crt2; target <- al
+  }
+
+  # j_temp <- ifelse(is.null(J), "", J)
+  # n_temp <- ifelse(is.null(n), "", n)
+  params <- list(delta = delta, delta_sd = delta_sd, rho = rho, rho_sd = rho_sd,
+                 rsq2 = rsq2, K = K, P = P, power = power, alpha = alpha)
+
 
   # Define a loss function for J or n, attempt to optimize using uniroot in the
   # specified internal, and try other optimization methods if root-finding fails.
   if (is.null(J)) { # solve for J
-    loss <- define_loss(solve_for_J = TRUE, squared = FALSE, ep, al, J = j_temp,
-                        n = n_temp, test = test, reparameterize = reparameterize,
-                        list_params = params)
+    loss <- function(J) {
+      do.call(criteria, append(list(J = J, n = n), params)) - target
+    }
+    # loss <- define_loss(solve_for_J = TRUE, squared = FALSE, ep, al, J = j_temp,
+    #                     n = n_temp, test = test,
+    #                     list_params = params)
 
     min_j <- if (is.null(al) & !is.null(ep)) (K + 2 + 1) else Jn_conv[1]
 
     J <- try(stats::uniroot(loss, interval = c(min_j, 1e8))$root, silent = TRUE)
 
     # if root-finding method fails, try optimization methods
     if (class(J) == "try-error") {
-      loss <- define_loss(solve_for_J = TRUE, squared = TRUE, ep, al,
-                          J = j_temp, n = n_temp, test = test,
-                          reparameterize = reparameterize, list_params = params)
+      loss <- function(J) {
+        (do.call(criteria, append(list(J = J, n = n), params)) - target)^2
+      }
+      # loss <- define_loss(solve_for_J = TRUE, squared = TRUE, ep, al,
+      #                     J = j_temp, n = n_temp, test = test,
+      #                     list_params = params)
       J <- Jn_optimize(start = min_j, loss = loss, lower = K + 3, upper = 1e6,
                        solve = "J")
     }
   } else { # solve n
-    loss <- define_loss(solve_for_J = FALSE, squared = FALSE, ep, al,
-                        J = j_temp, n = n_temp, test = test,
-                        reparameterize = reparameterize, list_params = params)
-    min_j <- 1
+    # loss <- define_loss(solve_for_J = FALSE, squared = FALSE, ep, al,
+    #                     J = j_temp, n = n_temp, test = test,
+    #                     list_params = params)
+    loss <- function(n) {
+      do.call(criteria, append(list(J = J, n = n), params)) - target
+    }
+    min_n <- 1
 
-    n <- try(stats::uniroot(loss, interval = c(min_j, 1e8))$root, silent = TRUE)
+    n <- try(stats::uniroot(loss, interval = c(min_n, 1e8))$root, silent = TRUE)
     if (class(n) == "try-error") {
-      loss <- define_loss(solve_for_J = FALSE, squared = TRUE, ep, al,
-                          J = j_temp, n = n_temp,test = test,
-                          reparameterize = reparameterize, list_params = params)
+      loss <- function(n) {
+        (do.call(criteria, append(list(J = J, n = n), params)) - target)^2
```

Comment:
@winniewytse It looks like we can delete the define_loss() helper function now, nice!